N^2 complexity in ChainedFilterSet regardless of how many actual conditions are passed

At the moment the initialization of ChainedFilterSet iterates over all possible combinations of fllters. Can this be changed to instead only iterate over filters that are passed as arguments?

    class ChainedFilterSet(django_filters.FilterSet):
        def __init__(self, *args, **kwargs):
        .....

        for name, filter_ in six.iteritems(self.filters):

Nope!

That breaks filtering for us:

```python
  File "rest_framework_chain/filterset.py", line 50, in __new__
    new_cls = super(ChainedFilterSet, cls).__new__(cls, *args, **kwargs)
TypeError: object() takes no parameters
```

After fixing that:

```python
======================================================================
FAIL: test_contact_ordering_and_filtering (api.tests.test_contact.ListContactTestCase)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "tests/test_contact.py", line 47, in test_contact_ordering_and_filtering
    'to_user__username__iexact': users[0].username.swapcase(),
  File "tests/drivers.py", line 79, in _test
    self.assertEqual(len(objects), len(results))
AssertionError: 2 != 6
```

I don't think it would help though since `__new__` is called each time an instance is created:

```python
>>> class Camelid():
...   def __new__(cls):
...     if cls.__name__ == 'Alpaca':
...       print("Yay! Alpacas!")
...     return super(Camelid, cls).__new__(cls)
... 
>>> class Alpaca(Camelid):
...   pass
... 
>>> Alpaca()
Yay! Alpacas!
<__main__.Alpaca object at 0x7ffc32c00668>
>>> Alpaca()
Yay! Alpacas!
<__main__.Alpaca object at 0x7ffc32c006a0>
```

I think what you're looking for is meta classes:

```python
>>> import six
>>> 
>>> class CamelidMeta(type):
...   def __new__(cls, name, bases, attrs):
...     if name == 'Alpaca':
...       print("Yay! Alpacas!")
...     return super(Camelid, cls).__new__(cls, name, bases, attrs)
... 
>>> class Alpaca(six.with_metaclass(CamelidMeta)):
...   pass
... 
Yay! Alpacas!
>>> Alpaca()
<__main__.Alpaca object at 0x7f10d0468f28>
>>> Alpaca()
<__main__.Alpaca object at 0x7f10ce3c78d0>
```

--- a/.travis.yml
+++ b/.travis.yml
@@ -4,6 +4,7 @@ python:
   - "3.2"
   - "3.3"
   - "3.4"
+  - "3.5"
 env:
   - DJANGO="Django>=1.4,<1.5"
   - DJANGO="Django>=1.5,<1.6"

--- a/.travis.yml
+++ b/.travis.yml
@@ -24,4 +24,12 @@ matrix:
       env: DJANGO="Django>=1.4,<1.5"
     - python: "3.4"
       env: DJANGO="Django>=1.4,<1.5"
+    - python: "3.5"
+      env: DJANGO="Django>=1.4,<1.5"
+    - python: "3.5"
+      env: DJANGO="Django>=1.5,<1.6"
+    - python: "3.5"
+      env: DJANGO="Django>=1.6,<1.7"
+    - python: "3.5"
+      env: DJANGO="Django>=1.7,<1.8"
   fast_finish: true

--- a/rest_framework_filters/backends.py
+++ b/rest_framework_filters/backends.py
@@ -1,50 +1,7 @@
-from copy import deepcopy
 
 import rest_framework.filters
-from rest_framework.compat import django_filters
-
 from .filterset import FilterSet
 
 
-
 class DjangoFilterBackend(rest_framework.filters.DjangoFilterBackend):
-    """
-    A version of `DjangoFilterBackend` that caches repeated filter classes
-    and uses our FilterSet by default.
-    """
     default_filter_set = FilterSet
-
-    _filter_instance_cache = {}
-
-    def _setup_filter_instance(self, _filter, query_params, queryset=None):
-        _filter.data = query_params or {}
-        _filter.is_bound = query_params is not None
-
-        if hasattr(_filter, '_qs'):
-            del _filter._qs
-        if hasattr(_filter, '_form'):
-            del _filter._form
-        if hasattr(_filter, '_ordering_field'):
-            del _filter._ordering_field
-
-        if queryset is None:
-            queryset = _filter._meta.model._default_manager.all()
-        _filter.queryset = queryset
-        # propagate the model being used through the filters
-        for f in _filter.filters.values():
-            f.model = _filter._meta.model
-
-    def filter_queryset(self, request, queryset, view):
-        filter_class = self.get_filter_class(view, queryset)
-
-        if filter_class:
-            if filter_class in self._filter_instance_cache:
-                _filter = self._filter_instance_cache[filter_class]
-                self._setup_filter_instance(_filter, request.query_params, queryset=queryset)
-            else:
-                _filter = filter_class(request.query_params, queryset=queryset)
-                self._filter_instance_cache[filter_class] = _filter
-
-            return _filter.qs
-
-        return queryset

--- a/rest_framework_filters/backends.py
+++ b/rest_framework_filters/backends.py
@@ -5,3 +5,13 @@ from .filterset import FilterSet
 
 class DjangoFilterBackend(rest_framework.filters.DjangoFilterBackend):
     default_filter_set = FilterSet
+    _related_filterset_cache = {}  # set to None to disable cache propagation
+
+    def filter_queryset(self, request, queryset, view):
+        filter_class = self.get_filter_class(view, queryset)
+
+        if filter_class:
+            cache = self._related_filterset_cache
+            return filter_class(request.query_params, queryset=queryset, cache=cache).qs
+
+        return queryset

--- a/rest_framework_filters/filters.py
+++ b/rest_framework_filters/filters.py
@@ -4,11 +4,17 @@ from __future__ import unicode_literals
 from django.utils import six
 
 from rest_framework.settings import api_settings
-import rest_framework.filters
 import django_filters
 from django_filters.filters import *
 
 from . import fields
+
+
+def _import_class(path):
+    module_path, class_name = path.rsplit('.', 1)
+    module = __import__(module_path, fromlist=[class_name], level=0)
+    return getattr(module, class_name)
+
 
 def subsitute_iso8601(date_type):
     from rest_framework import ISO_8601
@@ -42,18 +48,22 @@ DATETIME_INPUT_FORMATS = subsitute_iso8601('datetime')
 class RelatedFilter(ModelChoiceFilter):
     def __init__(self, filterset, *args, **kwargs):
         self.filterset = filterset
-        self.parent_relation = kwargs.get('parent_relation', None)
+        # self.parent_relation = kwargs.get('parent_relation', None)
         return super(RelatedFilter, self).__init__(*args, **kwargs)
 
-    def setup_filterset(self):
-        if isinstance(self.filterset, six.string_types):
-            # This is a recursive relation, defined via a string, so we need
-            # to create and import the class here.
-            items = self.filterset.split('.')
-            cls = str(items[-1]) # Ensure not unicode on py2.x
-            mod = __import__('.'.join(items[:-1]), fromlist=[cls])
-            self.filterset = getattr(mod, cls)
+    def filterset():
+        def fget(self):
+            if isinstance(self._filterset, six.string_types):
+                self._filterset = _import_class(self._filterset)
+            return self._filterset
 
+        def fset(self, value):
+            self._filterset = value
+
+        return locals()
+    filterset = property(**filterset())
+
+    def setup_filterset(self):
         self.extra['queryset'] = self.filterset._meta.model.objects.all()
 
 

--- a/rest_framework_filters/filters.py
+++ b/rest_framework_filters/filters.py
@@ -1,6 +1,7 @@
 from __future__ import absolute_import
 from __future__ import unicode_literals
 
+from collections import OrderedDict
 from django.utils import six
 
 from rest_framework.settings import api_settings
@@ -62,6 +63,27 @@ class RelatedFilter(ModelChoiceFilter):
 
         return locals()
     filterset = property(**filterset())
+
+    def get_filterset_subset(self, filter_names):
+        """
+        Returns a FilterSet subclass that contains the subset of filters
+        specified in `filter_names`. This is useful for creating FilterSets
+        used across relationships, as it minimizes the deepcopy overhead
+        incurred when instantiating the FilterSet.
+        """
+        BaseFilterSet = self.filterset
+
+        class FilterSetSubset(BaseFilterSet):
+            pass
+
+        FilterSetSubset.__name__ = str('%sSubset' % (BaseFilterSet.__name__))
+        FilterSetSubset.base_filters = OrderedDict([
+            (name, f)
+            for name, f in six.iteritems(BaseFilterSet.base_filters)
+            if name in filter_names
+        ])
+
+        return FilterSetSubset
 
     def setup_filterset(self):
         self.extra['queryset'] = self.filterset._meta.model.objects.all()

--- a/rest_framework_filters/filterset.py
+++ b/rest_framework_filters/filterset.py
@@ -69,6 +69,8 @@ class FilterSet(six.with_metaclass(FilterSetMetaclass, filterset.FilterSet)):
     }
 
     def __init__(self, *args, **kwargs):
+        self._related_filterset_cache = kwargs.pop('cache', {})
+
         super(FilterSet, self).__init__(*args, **kwargs)
 
         for name, filter_ in six.iteritems(self.filters):
@@ -86,14 +88,14 @@ class FilterSet(six.with_metaclass(FilterSetMetaclass, filterset.FilterSet)):
         """
         requested_filters = OrderedDict()
 
-        # filter out any filters not included in the request data
+        # Add plain lookup filters if match. ie, `username__icontains`
         for filter_key, filter_value in six.iteritems(self.filters):
             if filter_key in self.data:
                 requested_filters[filter_key] = filter_value
 
-        # build a map of potential {rel: [filter]} pairs
+        # build a map of potential {rel: {filter: value}} data
         related_data = OrderedDict()
-        for filter_key in self.data:
+        for filter_key, value in six.iteritems(self.data):
             if filter_key not in self.filters:
 
                 # skip non lookup/related keys
@@ -102,11 +104,11 @@ class FilterSet(six.with_metaclass(FilterSetMetaclass, filterset.FilterSet)):
 
                 rel_name, filter_key = filter_key.split(LOOKUP_SEP, 1)
 
-                related_data.setdefault(rel_name, [])
-                related_data[rel_name].append(filter_key)
+                related_data.setdefault(rel_name, OrderedDict())
+                related_data[rel_name][filter_key] = value
 
-        # walk the related lookup data. If the rel is a RelatedFilter,
-        # then instantiate its filterset and append its filters
+        # walk the related lookup data. If the filter is a RelatedFilter,
+        # then instantiate its filterset and append its filters.
         for rel_name, rel_data in related_data.items():
             related_filter = self.filters.get(rel_name, None)
 
@@ -114,15 +116,53 @@ class FilterSet(six.with_metaclass(FilterSetMetaclass, filterset.FilterSet)):
             if not isinstance(related_filter, filters.RelatedFilter):
                 continue
 
-            filterset = related_filter.filterset(data=rel_data)
-            rel_filters = filterset.get_filters()
+            # get known filter names
+            filterset_class = related_filter.filterset
+            filter_names = [filterset_class.get_filter_name(param) for param in rel_data.keys()]
 
+            # attempt to retrieve related filterset subset from the cache
+            key = self.cache_key(filterset_class, filter_names)
+            subset_class = self.cache_get(key)
+
+            # otherwise build and insert it into the cache
+            if subset_class is None:
+                subset_class = related_filter.get_filterset_subset(filter_names)
+                self.cache_set(key, subset_class)
+
+            # initialize and copy filters
+            filterset = subset_class(data=rel_data)
+            rel_filters = filterset.get_filters()
             for filter_key, filter_value in six.iteritems(rel_filters):
+                # modify filter name to account for relationship
                 rel_filter_key = LOOKUP_SEP.join([rel_name, filter_key])
                 filter_value.name = LOOKUP_SEP.join([related_filter.name, filter_value.name])
                 requested_filters[rel_filter_key] = filter_value
 
         return requested_filters
+
+    @classmethod
+    def get_filter_name(cls, param):
+        """
+        Get the filter name for the request data parameter.
+        """
+        # Attempt to match against filters with lookups first. (username__endswith)
+        if param in cls.base_filters:
+            return param
+
+        # Fallback to matching against relationships. (author__username__endswith)
+        param = param.split(LOOKUP_SEP, 1)[0]
+        f = cls.base_filters.get(param, None)
+        if isinstance(f, filters.RelatedFilter):
+            return param
+
+    def cache_key(self, filterset, filter_names):
+        return '%sSubset-%s' % (filterset.__name__, '-'.join(sorted(filter_names)), )
+
+    def cache_get(self, key):
+        return self._related_filterset_cache.get(key)
+
+    def cache_set(self, key, value):
+        self._related_filterset_cache[key] = value
 
     @property
     def qs(self):

--- a/rest_framework_filters/filterset.py
+++ b/rest_framework_filters/filterset.py
@@ -1,7 +1,6 @@
 from __future__ import absolute_import
 from __future__ import unicode_literals
 
-from copy import copy
 from collections import OrderedDict
 
 try:
@@ -19,12 +18,42 @@ from django.utils import six
 
 import django_filters
 import django_filters.filters
-from django_filters.filterset import get_model_field
+from django_filters import filterset
 
 from . import filters
 
 
-class FilterSet(django_filters.FilterSet):
+class FilterSetMetaclass(filterset.FilterSetMetaclass):
+    def __new__(cls, name, bases, attrs):
+        new_class = super(FilterSetMetaclass, cls).__new__(cls, name, bases, attrs)
+
+        # Populate our FilterSet fields with all the possible
+        # filters for the AllLookupsFilter field.
+        for name, filter_ in six.iteritems(new_class.base_filters):
+            if isinstance(filter_, filters.AllLookupsFilter):
+                model = new_class._meta.model
+                field = filterset.get_model_field(model, filter_.name)
+
+                for lookup_type in django_filters.filters.LOOKUP_TYPES:
+                    if isinstance(field, ForeignObjectRel):
+                        f = new_class.filter_for_reverse_field(field, filter_.name)
+                    else:
+                        f = new_class.filter_for_field(field, filter_.name)
+                    f.lookup_type = lookup_type
+                    f = new_class.fix_filter_field(f)
+
+                    # compute filter name
+                    filter_name = name
+                    # Don't add "exact" to filter names
+                    if lookup_type != 'exact':
+                        filter_name = LOOKUP_SEP.join([name, lookup_type])
+
+                    new_class.base_filters[filter_name] = f
+
+        return new_class
+
+
+class FilterSet(six.with_metaclass(FilterSetMetaclass, filterset.FilterSet)):
     # In order to support ISO-8601 -- which is the default output for
     # DRF -- we need to set up custom date/time input formats.
     filter_overrides = {
@@ -39,8 +68,6 @@ class FilterSet(django_filters.FilterSet):
         },
     }
 
-    LOOKUP_TYPES = django_filters.filters.LOOKUP_TYPES
-
     def __init__(self, *args, **kwargs):
         super(FilterSet, self).__init__(*args, **kwargs)
 
@@ -51,20 +78,6 @@ class FilterSet(django_filters.FilterSet):
                 # Add an 'isnull' filter to allow checking if the relation is empty.
                 isnull_filter = filters.BooleanFilter(name=("%s%sisnull" % (filter_.name, LOOKUP_SEP)))
                 self.filters['%s%s%s' % (filter_.name, LOOKUP_SEP, 'isnull')] = isnull_filter
-
-            elif isinstance(filter_, filters.AllLookupsFilter):
-                # Populate our FilterSet fields with all the possible
-                # filters for the AllLookupsFilter field.
-                model = self._meta.model
-                field = get_model_field(model, filter_.name)
-                for lookup_type in self.LOOKUP_TYPES:
-                    if isinstance(field, ForeignObjectRel):
-                        f = self.filter_for_reverse_field(field, filter_.name)
-                    else:
-                        f = self.filter_for_field(field, filter_.name)
-                    f.lookup_type = lookup_type
-                    f = self.fix_filter_field(f)
-                    self.filters["%s%s%s" % (name, LOOKUP_SEP, lookup_type)] = f
 
     def get_filters(self):
         """
@@ -122,9 +135,10 @@ class FilterSet(django_filters.FilterSet):
 
         return qs
 
-    def fix_filter_field(self, f):
+    @classmethod
+    def fix_filter_field(cls, f):
         """
-        Fix the filter field based on the lookup type. 
+        Fix the filter field based on the lookup type.
         """
         lookup_type = f.lookup_type
         if lookup_type == 'isnull':

--- a/rest_framework_filters/filterset.py
+++ b/rest_framework_filters/filterset.py
@@ -29,7 +29,7 @@ class FilterSetMetaclass(filterset.FilterSetMetaclass):
 
         # Populate our FilterSet fields with all the possible
         # filters for the AllLookupsFilter field.
-        for name, filter_ in six.iteritems(new_class.base_filters):
+        for name, filter_ in six.iteritems(new_class.base_filters.copy()):
             if isinstance(filter_, filters.AllLookupsFilter):
                 model = new_class._meta.model
                 field = filterset.get_model_field(model, filter_.name)
@@ -74,7 +74,7 @@ class FilterSet(six.with_metaclass(FilterSetMetaclass, filterset.FilterSet)):
 
         super(FilterSet, self).__init__(*args, **kwargs)
 
-        for name, filter_ in six.iteritems(self.filters):
+        for name, filter_ in six.iteritems(self.filters.copy()):
             if isinstance(filter_, filters.RelatedFilter):
                 filter_.setup_filterset()
 

--- a/rest_framework_filters/filterset.py
+++ b/rest_framework_filters/filterset.py
@@ -2,6 +2,7 @@ from __future__ import absolute_import
 from __future__ import unicode_literals
 
 from copy import copy
+from collections import OrderedDict
 
 try:
     from django.db.models.constants import LOOKUP_SEP
@@ -45,13 +46,12 @@ class FilterSet(django_filters.FilterSet):
 
         for name, filter_ in six.iteritems(self.filters):
             if isinstance(filter_, filters.RelatedFilter):
-                # Populate our FilterSet fields with the fields we've stored
-                # in RelatedFilter.
                 filter_.setup_filterset()
-                self.populate_from_filterset(filter_.filterset, filter_, name)
+
                 # Add an 'isnull' filter to allow checking if the relation is empty.
                 isnull_filter = filters.BooleanFilter(name=("%s%sisnull" % (filter_.name, LOOKUP_SEP)))
                 self.filters['%s%s%s' % (filter_.name, LOOKUP_SEP, 'isnull')] = isnull_filter
+
             elif isinstance(filter_, filters.AllLookupsFilter):
                 # Populate our FilterSet fields with all the possible
                 # filters for the AllLookupsFilter field.
@@ -66,6 +66,62 @@ class FilterSet(django_filters.FilterSet):
                     f = self.fix_filter_field(f)
                     self.filters["%s%s%s" % (name, LOOKUP_SEP, lookup_type)] = f
 
+    def get_filters(self):
+        """
+        Build a set of filters based on the requested data. The resulting set
+        will walk `RelatedFilter`s to recursively build the set of filters.
+        """
+        requested_filters = OrderedDict()
+
+        # filter out any filters not included in the request data
+        for filter_key, filter_value in six.iteritems(self.filters):
+            if filter_key in self.data:
+                requested_filters[filter_key] = filter_value
+
+        # build a map of potential {rel: [filter]} pairs
+        related_data = OrderedDict()
+        for filter_key in self.data:
+            if filter_key not in self.filters:
+
+                # skip non lookup/related keys
+                if LOOKUP_SEP not in filter_key:
+                    continue
+
+                rel_name, filter_key = filter_key.split(LOOKUP_SEP, 1)
+
+                related_data.setdefault(rel_name, [])
+                related_data[rel_name].append(filter_key)
+
+        # walk the related lookup data. If the rel is a RelatedFilter,
+        # then instantiate its filterset and append its filters
+        for rel_name, rel_data in related_data.items():
+            related_filter = self.filters.get(rel_name, None)
+
+            # skip non-`RelatedFilter`s
+            if not isinstance(related_filter, filters.RelatedFilter):
+                continue
+
+            filterset = related_filter.filterset(data=rel_data)
+            rel_filters = filterset.get_filters()
+
+            for filter_key, filter_value in six.iteritems(rel_filters):
+                rel_filter_key = LOOKUP_SEP.join([rel_name, filter_key])
+                filter_value.name = LOOKUP_SEP.join([related_filter.name, filter_value.name])
+                requested_filters[rel_filter_key] = filter_value
+
+        return requested_filters
+
+    @property
+    def qs(self):
+        available_filters = self.filters
+        requested_filters = self.get_filters()
+
+        self.filters = requested_filters
+        qs = super(FilterSet, self).qs
+        self.filters = available_filters
+
+        return qs
+
     def fix_filter_field(self, f):
         """
         Fix the filter field based on the lookup type. 
@@ -76,38 +132,3 @@ class FilterSet(django_filters.FilterSet):
         if lookup_type == 'in' and type(f) in [filters.NumberFilter]:
             return filters.InSetNumberFilter(name=("%s%sin" % (f.name, LOOKUP_SEP)))
         return f
-
-    def populate_from_filterset(self, filterset, filter_, name):
-        """
-        Populate `filters` with filters provided on `filterset`.
-        """
-        def _should_skip():
-            for name, filter_ in six.iteritems(self.filters):
-                if filter_value == filter_:
-                    return True
-                # Avoid infinite recursion on recursive relations.  If the queryset and
-                # class are the same, then we assume that we've already added this
-                # filter previously along the lookup chain, e.g.
-                # a__b__a <-- the last 'a' there.
-                if (isinstance(filter_, filters.RelatedFilter) and
-                    isinstance(filter_value, filters.RelatedFilter)):
-                    if filter_value.extra.get('queryset', None) == filter_.extra.get('queryset'):
-                        return True
-            return False
-
-        for (filter_key, filter_value) in filterset.base_filters.items():
-            if _should_skip():
-                continue
-
-            filter_value = copy(filter_value)
-
-            # Guess on the field to join on, if applicable
-            if not getattr(filter_value, 'parent_relation', None):
-                filter_value.parent_relation = filterset._meta.model.__name__.lower()
-
-            # We use filter_.name -- which is the internal name, to do the actual query
-            filter_name = filter_value.name
-            filter_value.name = '%s%s%s' % (filter_.name, LOOKUP_SEP, filter_name)
-            # and then we use the /given/ name keyword as the actual querystring lookup, and
-            # the filter's name in the related class (filter_key).
-            self.filters['%s%s%s' % (name, LOOKUP_SEP, filter_key)] = filter_value

--- a/rest_framework_filters/tests.py
+++ b/rest_framework_filters/tests.py
@@ -669,3 +669,14 @@ class TestFilterSets(TestCase):
         self.assertEqual(len(f), 2)
         self.assertIn(p1, f)
         self.assertIn(p2, f)
+
+    def test_get_filterset_subset(self):
+        related_filter = NoteFilterWithRelated.base_filters['author']
+        filterset_class = related_filter.get_filterset_subset(['email'])
+
+        # ensure that the class name is useful when debugging
+        self.assertEqual(filterset_class.__name__, 'UserFilterSubset')
+
+        # ensure that the FilterSet subset only contains the requested fields
+        self.assertIn('email', filterset_class.base_filters)
+        self.assertEqual(len(filterset_class.base_filters), 1)

